hazelcast/src/main/java/com/hazelcast/jet/impl/LightMasterContext.java

```diff
@@ -193,31 +198,49 @@ private void finalizeJob(@Nullable final Throwable failure) {
         CompletableFuture<Void> combined = CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]));
 
         combined.whenComplete((ignored, e) -> {
-            Throwable fail = failure;
-            if (fail == null) {
-                jobCompletionFuture.complete(null);
-                jobEventService.publishEvent(jobId, RUNNING, COMPLETED, null, false);
-            } else {
-                TerminationMode requestedTerminationMode = fail instanceof JobTerminateRequestedException
-                        ? ((JobTerminateRequestedException) fail).mode() : null;
-                // translate JobTerminateRequestedException(CANCEL_FORCEFUL)
-                // to CancellationException or CancellationByUserException
-                if (requestedTerminationMode == CANCEL_FORCEFUL) {
-                    Throwable newFailure = userInitiatedTermination
-                            ? new CancellationByUserException()
-                            : new CancellationException();
-                    newFailure.initCause(failure);
-                    fail = newFailure;
+            lock.lock();
```

Comment:
I wonder if this `lock` adds measureable overhead, especially for simple, fast SQL queries

Reply:
I didn't measure the impact but AFAIK if the same thread constantly acquires/releases the lock, the impact is very minimal.

Reply:
we acquire the lock each time we finalize the job. Do you mean that if the lock is not contented (no concurrent listener registration), then the overhead and latency increase should be negligible, right?